File: theorem-context-sdk/theorem-context-py/theorem_context/product.py
"""Async client for the standalone THG product service."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

from .errors import (
    AuthError,
    HarnessError,
    RequestTimeoutError,
    ServerUnavailableError,
)
from .types import THGResult


def _quote_path(value: str) -> str:
    return quote(str(value), safe='')


class TheoremHotGraphClient:
    """Tenant-scoped client for the THG product service."""

    def __init__(
        self,
        *,
        base_url: str,
        token: str,
        tenant_id: str,
        timeout: float = 30.0,
        transport: httpx.AsyncBaseTransport | None = None,
    ) -> None:
        self.base_url = base_url.rstrip('/')
        self.token = token
        self.tenant_id = tenant_id
        self._http = httpx.AsyncClient(timeout=timeout, transport=transport)

    async def __aenter__(self) -> 'TheoremHotGraphClient':
        return self

    async def __aexit__(self, *_args: Any) -> None:
        await self.aclose()

    async def aclose(self) -> None:
        await self._http.aclose()

    async def command(
        self,
        command: str,
        args: dict[str, Any] | None = None,
    ) -> THGResult:
        return await self._post('/command', {
            'command': command,
            'args': args or {},
        })

    async def batch(
        self,
        commands: list[dict[str, Any]],
    ) -> dict[str, Any]:
        response = await self._request(
            'POST',
            '/batch',
            action='batch',
            json={'commands': commands},
        )
        return response.json()

    async def run(self, run_id: str) -> THGResult:
        response = await self._request(
            'GET',
            f'/runs/{_quote_path(run_id)}',
            action='run',
        )
        return THGResult.model_validate(response.json())

    async def context_pack(self, args: dict[str, Any]) -> THGResult:
        return await self._post('/context/pack', args)

    async def graph_query(
        self,
        query: str,
        *,
        graph: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> THGResult:
        return await self._post('/graph/query', {
            'query': query,
            'graph': graph or {},
            'params': params or {},
        })

    async def instant_kg_status(
        self,
        *,
        manifest: dict[str, Any] | None = None,
        delta: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return await self._post_dict('/instant-kg/status', {
            'manifest': manifest,
            'delta': delta,
        })

    async def instant_kg_ppr(
        self,
        seeds: dict[str, float],
        *,
        manifest: dict[str, Any] | None = None,
        delta: dict[str, Any] | None = None,
        alpha: float = 0.15,
        epsilon: float = 1e-4,
        max_pushes: int = 200_000,
        top_k: int = 10,
    ) -> dict[str, Any]:
        return await self._post_dict('/instant-kg/ppr', {
            'manifest': manifest,
            'delta': delta,
            'seeds': seeds,
            'alpha': alpha,
            'epsilon': epsilon,
            'max_pushes': max_pushes,
            'top_k': top_k,
        })

    async def instant_kg_impact(
        self,
        *,
        seed: str | None = None,
        symbol_name: str | None = None,
        manifest: dict[str, Any] | None = None,
        delta: dict[str, Any] | None = None,
        direction: str = 'out',
        max_depth: int = 2,
    ) -> dict[str, Any]:
        return await self._post_dict('/instant-kg/impact', {
            'manifest': manifest,
            'delta': delta,
            'seed': seed,
            'symbol_name': symbol_name,
            'direction': direction,
            'max_depth': max_depth,
        })

    async def instant_kg_related_objects(
        self,
        seed: str,
        *,
        manifest: dict[str, Any] | None = None,
        delta: dict[str, Any] | None = None,
        kinds: list[str] | None = None,
        top_k: int = 10,
    ) -> dict[str, Any]:
        return await self._post_dict('/instant-kg/related-objects', {
            'manifest': manifest,
            'delta': delta,
            'seed': seed,
            'kinds': kinds or [],
            'top_k': top_k,
        })

    async def instant_kg_search(
        self,
        query: str,
        *,
        manifest: dict[str, Any] | None = None,
        delta: dict[str, Any] | None = None,
        kinds: list[str] | None = None,
        top_k: int = 10,
    ) -> dict[str, Any]:
        return await self._post_dict('/instant-kg/search', {
            'manifest': manifest,
            'delta': delta,
            'query': query,
            'kinds': kinds or [],
            'top_k': top_k,
        })

    async def instant_kg_explain_edge(
        self,
        src: str,
        dst: str,
        *,
        manifest: dict[str, Any] | None = None,
        delta: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return await self._post_dict('/instant-kg/explain-edge', {
            'manifest': manifest,
            'delta': delta,
            'src': src,
            'dst': dst,
        })

    async def _post(self, path: str, body: dict[str, Any]) -> THGResult:
        response = await self._request(
            'POST',
            path,
            action=path,
            json=body,
        )
        return THGResult.model_validate(response.json())

    async def _post_dict(self, path: str, body: dict[str, Any]) -> dict[str, Any]:
        response = await self._request(
            'POST',
            path,
            action=path,
            json={key: value for key, value in body.items() if value is not None},
        )
        return response.json()

    async def _request(
        self,
        method: str,
        path: str,
        *,
        action: str,
        **kwargs: Any,
    ) -> httpx.Response:
        try:
            response = await self._http.request(
                method,
                f'{self._tenant_url()}{path}',
                headers=self._headers(),
                **kwargs,
            )
        except httpx.TimeoutException as exc:
            raise RequestTimeoutError(
                f'THG product {action} failed: {exc}',
            ) from exc
        except httpx.RequestError as exc:
            raise ServerUnavailableError(
                f'THG product {action} failed: {exc}',
            ) from exc
        self._raise_for_status(response, action=action)
        return response

    def _raise_for_status(self, response: httpx.Response, *, action: str) -> None:
        if response.is_success:
            return
        detail = response.text.strip()
        suffix = f' {detail}' if detail else ''
        message = (
            f'THG product {action} failed with HTTP '
            f'{response.status_code}{suffix}'
        )
        if response.status_code in {401, 403}:
            raise AuthError(message)
        if response.status_code in {408, 504}:
            raise RequestTimeoutError(message)
        if response.status_code in {502, 503}:
            raise ServerUnavailableError(message)
        raise HarnessError(message)

    def _headers(self) -> dict[str, str]:
        return {
            'Authorization': f'Bearer {self.token}',
            'Content-Type': 'application/json',
        }

    def _tenant_url(self) -> str:
        return f'{self.base_url}/v1/tenants/{_quote_path(self.tenant_id)}'

File: theorem-context-sdk/theorem-context-ts/src/product.ts
import type { THGResult } from './types.js';
import {
  AuthError,
  HarnessError,
  RequestTimeoutError,
  ServerUnavailableError,
} from './errors.js';

export interface TheoremHotGraphClientOptions {
  baseUrl: string;
  token: string;
  tenantId: string;
  fetchImpl?: typeof fetch;
}

export type THGProductResult = THGResult & {
  ok: boolean;
  error?: Record<string, unknown> | null;
};

export type InstantKgPayload = {
  manifest?: Record<string, unknown>;
  delta?: Record<string, unknown>;
};

export class TheoremHotGraphClient {
  private readonly baseUrl: string;
  private readonly token: string;
  private readonly tenantId: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: TheoremHotGraphClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.token = options.token;
    this.tenantId = options.tenantId;
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  async command(
    command: string,
    args: Record<string, unknown> = {},
  ): Promise<THGProductResult> {
    return this.post('/command', { command, args });
  }

  async batch(
    commands: Array<{ command: string; args?: Record<string, unknown> }>,
  ): Promise<{ ok: boolean; results: THGProductResult[]; state_hash: string }> {
    return this.post('/batch', { commands });
  }

  async run(runId: string): Promise<THGProductResult> {
    const response = await this.request(
      `${this.tenantUrl()}/runs/${encodeURIComponent(runId)}`,
      { method: 'GET', headers: this.headers() },
      'run',
    );
    return this.parseResponse(response);
  }

  async contextPack(args: Record<string, unknown>): Promise<THGProductResult> {
    return this.post('/context/pack', args);
  }

  async graphQuery(
    query: string,
    options: {
      graph?: Record<string, unknown>;
      params?: Record<string, unknown>;
    } = {},
  ): Promise<THGProductResult> {
    return this.post('/graph/query', {
      query,
      graph: options.graph ?? {},
      params: options.params ?? {},
    });
  }

  async instantKgStatus(options: InstantKgPayload = {}): Promise<Record<string, unknown>> {
    return this.post('/instant-kg/status', this.cleanBody(options));
  }

  async instantKgPpr(
    seeds: Record<string, number>,
    options: InstantKgPayload & {
      alpha?: number;
      epsilon?: number;
      maxPushes?: number;
      topK?: number;
    } = {},
  ): Promise<Record<string, unknown>> {
    return this.post('/instant-kg/ppr', this.cleanBody({
      manifest: options.manifest,
      delta: options.delta,
      seeds,
      alpha: options.alpha ?? 0.15,
      epsilon: options.epsilon ?? 0.0001,
      max_pushes: options.maxPushes ?? 200000,
      top_k: options.topK ?? 10,
    }));
  }

  async instantKgImpact(options: InstantKgPayload & {
    seed?: string;
    symbolName?: string;
    direction?: 'out' | 'in';
    maxDepth?: number;
  }): Promise<Record<string, unknown>> {
    return this.post('/instant-kg/impact', this.cleanBody({
      manifest: options.manifest,
      delta: options.delta,
      seed: options.seed,
      symbol_name: options.symbolName,
      direction: options.direction ?? 'out',
      max_depth: options.maxDepth ?? 2,
    }));
  }

  async instantKgRelatedObjects(
    seed: string,
    options: InstantKgPayload & {
      kinds?: string[];
      topK?: number;
    } = {},
  ): Promise<Record<string, unknown>> {
    return this.post('/instant-kg/related-objects', this.cleanBody({
      manifest: options.manifest,
      delta: options.delta,
      seed,
      kinds: options.kinds ?? [],
      top_k: options.topK ?? 10,
    }));
  }

  async instantKgSearch(
    query: string,
    options: InstantKgPayload & {
      kinds?: string[];
      topK?: number;
    } = {},
  ): Promise<Record<string, unknown>> {
    return this.post('/instant-kg/search', this.cleanBody({
      manifest: options.manifest,
      delta: options.delta,
      query,
      kinds: options.kinds ?? [],
      top_k: options.topK ?? 10,
    }));
  }

  async instantKgExplainEdge(
    src: string,
    dst: string,
    options: InstantKgPayload = {},
  ): Promise<Record<string, unknown>> {
    return this.post('/instant-kg/explain-edge', this.cleanBody({
      manifest: options.manifest,
      delta: options.delta,
      src,
      dst,
    }));
  }

  private async post<T>(path: string, body: Record<string, unknown>): Promise<T> {
    const response = await this.request(
      `${this.tenantUrl()}${path}`,
      {
        method: 'POST',
        headers: this.headers(),
        body: JSON.stringify(body),
      },
      path,
    );
    return this.parseResponse(response) as Promise<T>;
  }

  private async request(
    url: string,
    init: RequestInit,
    action: string,
  ): Promise<Response> {
    let response: Response;
    try {
      response = await this.fetchImpl(url, init);
    } catch (error) {
      throw mapProductTransportError(action, error);
    }
    if (!response.ok) {
      throw await mapProductHttpError(action, response);
    }
    return response;
  }

  private async parseResponse<T>(response: Response): Promise<T> {
    return (await response.json()) as T;
  }

  private headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.token}`,
      'Content-Type': 'application/json',
    };
  }

  private tenantUrl(): string {
    return `${this.baseUrl}/v1/tenants/${encodeURIComponent(this.tenantId)}`;
  }

  private cleanBody(body: Record<string, unknown>): Record<string, unknown> {
    return Object.fromEntries(
      Object.entries(body).filter(([, value]) => value !== undefined && value !== null),
    );
  }
}

async function mapProductHttpError(
  action: string,
  response: Response,
): Promise<Error> {
  const detail = (await safeResponseText(response)).trim();
  const suffix = detail ? ` ${detail}` : '';
  const message = `THG product ${action} failed with HTTP ${response.status}${suffix}`;
  if (response.status === 401 || response.status === 403) {
    return new AuthError(message);
  }
  if (response.status === 408 || response.status === 504) {
    return new RequestTimeoutError(message);
  }
  if (response.status === 502 || response.status === 503) {
    return new ServerUnavailableError(message);
  }
  return new HarnessError(message);
}

function mapProductTransportError(action: string, error: unknown): Error {
  const message = error instanceof Error ? error.message : 'unknown error';
  if (error instanceof Error && isTimeoutError(error)) {
    return new RequestTimeoutError(`THG product ${action} failed: ${message}`);
  }
  if (error instanceof Error && isServerUnavailableError(error)) {
    return new ServerUnavailableError(`THG product ${action} failed: ${message}`);
  }
  return new HarnessError(`THG product ${action} failed: ${message}`);
}

async function safeResponseText(response: Response): Promise<string> {
  try {
    return await response.text();
  } catch {
    return '';
  }
}

function isTimeoutError(error: Error): boolean {
  return error.name === 'AbortError' || /timed?\s*out/i.test(error.message);
}

function isServerUnavailableError(error: Error): boolean {
  return (
    error.name === 'TypeError'
    || /ECONNREFUSED|ECONNRESET|ENOTFOUND|EAI_AGAIN|network/i.test(error.message)
  );
}
